src/Compiler/Optimize/LowerComputedCollections.fs

```diff
@@ -399,7 +416,64 @@ module Array =
                     )
             )
 
-let LowerComputedListOrArrayExpr tcVal (g: TcGlobals) amap overallExpr =
+/// f (); …; Seq.singleton x
+///
+/// E.g., in [for x in … do f (); …; yield x]
+[<return: Struct>]
+let (|SimpleSequential|_|) g expr =
+    let rec loop expr cont =
+        match expr with
+        | Expr.Sequential (expr1, DebugPoints (ValApp g g.seq_singleton_vref (_, [body], _), debug), kind, m) ->
+            ValueSome (cont (expr1, debug body, kind, m))
+
+        | Expr.Sequential (expr1, body, kind, m) ->
+            loop body (cont >> fun body -> Expr.Sequential (expr1, body, kind, m))
+
+        | _ -> ValueNone
+
+    loop expr Expr.Sequential
+
+/// The representation used for
+///
+/// for … in … -> …
+///
+/// and
+///
+/// for … in … do yield …
+[<return: Struct>]
+let (|SeqMap|_|) g expr =
+    match expr with
+    | ValApp g g.seq_map_vref ([ty1; ty2], [Expr.Lambda (valParams = [loopVal]; bodyExpr = body) as mapping; input], _) ->
+        ValueSome (ty1, ty2, input, mapping, loopVal, body)
+    | _ -> ValueNone
+
+/// The representation used for
+///
+/// for … in … do f (); …; yield …
+[<return: Struct>]
+let (|SeqCollectSingle|_|) g expr =
+    match expr with
+    | ValApp g g.seq_collect_vref ([ty1; _; ty2], [Expr.Lambda (valParams = [loopVal]; bodyExpr = SimpleSequential g body) as mapping; input], _) ->
+        ValueSome (ty1, ty2, input, mapping, loopVal, body)
+    | _ -> ValueNone
+
+/// for … in … -> …
+/// for … in … do yield …
+/// for … in … do f (); …; yield …
+[<return: Struct>]
+let (|SimpleMapping|_|) g expr =
+    match expr with
+    // for … in … -> …
+    // for … in … do yield …
+    | ValApp g g.seq_delay_vref (_, [Expr.Lambda (bodyExpr = SeqMap g (ty1, ty2, input, mapping, loopVal, body))], _)
```

Comment:
Can we guarantee this works the same w.r.t. order of side-effects inside `f()` (probably yes) as well as possible exceptions thrown from `f()` ?

I assume that all usages where body contains either `try-with` or `try-finally` are out of scope, so also yes?


(If there was a potential for a different behavior, we would for sure need a language feature for this)

Reply:
Hmm, I can't think of how there would be an observable difference (but let me know if you can).

We're basically going from

```csharp
var list = new List<int>();
// Compiles to a try/finally and a while (enumerator.MoveNext()), etc.
foreach (var n in Enumerable.Range(1, 10))
{
    list.Add(F(n));
}
list.ToArray();
```

to

```csharp
var array = new int[10];
var n = 1;
for (var i = 0; i < 10; i++)
{
    array[i] = F(n);
    n += step;
    i++;
}
```

This is behind the same language feature flag as #16650, fwiw (since it's just an extension of that).

Reply:
> order of side-effects inside `f()`

Did you mean inside `f` itself (which this definitely shouldn't affect), or in something like this involving multiple sequential expressions?

https://github.com/dotnet/fsharp/blob/5da1a7c18820ecf3502d4741ec226aca0ab5857a/tests/FSharp.Compiler.ComponentTests/EmittedIL/ComputedCollections/ForNInRangeArrays.fs#L2

In this case, order is kept — you can see that arg 0 (`f`) is loaded and called before arg 1 (`g`) in the body of the loop:

https://github.com/dotnet/fsharp/blob/5da1a7c18820ecf3502d4741ec226aca0ab5857a/tests/FSharp.Compiler.ComponentTests/EmittedIL/ComputedCollections/ForNInRangeArrays.fs.il.bsl#L95-L128

Reply:
...Hmm, technically a version of one of the scenarios outlined in https://github.com/dotnet/fsharp/pull/16832#discussion_r1520432627 applies to this PR as well (i.e., even with exactly one `yield`).

This is the scenario:

1. A computed array expression is being initialized from a `for … in … -> …` over an integral range.
2. The integral type is 64-bit.
3. The total count is greater than `System.Array.MaxLength` ($2,147,483,591$).
4. The `yield` expression or pattern binding has a side-effect.

#### Examples

```fsharp
let f n = printfn $"{n}"; n
[|for n in UInt32.MinValue..UInt32.MaxValue -> f n|]
```

```fsharp
let (|SideEffect|) n = printfn $"{n}"; n
[|for SideEffect n in UInt32.MinValue..UInt32.MaxValue -> n|]
```

#### Behavior before this PR

- `n` gets printed until an overflow exception is thrown when `ArrayCollector` tries to resize the backing array to something bigger than `System.Array.MaxLength`.

#### Behavior after this PR

- `n` never gets printed, since an overflow exception is thrown when we try to initialize the array at the very beginning with $count = 2^{32} + 1$.

An overflow exception is thrown both before and after; the only observable difference is that, before, an unspecified number of side-effects may be executed before the overflow exception is raised, whereas no side-effects will be executed after this PR.

I would argue that it is fine to make this runtime change:

- No legitimate code actually relies on this.
- Any code that _does_ rely on the existing behavior in this exact scenario is inherently unsafe and makes no sense.
- The number of iterations, if any, before an overflow exception is thrown in this scenario is not specified, so no one should be relying on a specific implementation's behavior.
- This change is under a language version flag. If someone actually has such code and wants to keep it behaving as before, they can target F# 8.0 or lower.

Reply:
I agree to the reasoning at the end.
The fact that previous implementation was using doubling strategy and not upfront allocation was an implementation detail.